library/src/main/java/com/novoda/downloadmanager/Batch.java

```diff
@@ -61,40 +61,38 @@ public String toString() {
         return "Batch{"
                 + "downloadBatchId=" + downloadBatchId
                 + ", title='" + title + '\''
-                + ", networkAddressAndFilePathById=" + networkAddressAndFilePathById
+                + ", files=" + files
                 + '}';
     }
 
     public static class Builder {
 
         private final DownloadBatchId downloadBatchId;
         private final String title;
-        private final Map<DownloadFileId, NetworkAddressAndFilePath> networkAddressAndFilePathById = new HashMap<>();
+        private List<File> files;
 
         public Builder(DownloadBatchId downloadBatchId, String title) {
             this.downloadBatchId = downloadBatchId;
             this.title = title;
+            files = new ArrayList<>();
         }
 
-        public Builder addFile(String fileUrl) {
-            String rawId = downloadBatchId.rawId() + fileUrl;
-            DownloadFileId downloadFileId = DownloadFileIdCreator.createFrom(rawId);
-            return addFile(downloadFileId, fileUrl);
+        Builder withFile(File file) {
```

Comment:
This is probably my biggest beef with this pattern, wondering if it would be better to introduce two different interfaces to better define these `Builder` methods. 

`withFile` is called from the `File.Builder` when the `apply` is called. The `apply` builds the final `File` object and then applies it to the parent builder with this `withFile` method.